### PR TITLE
[ALLUXIO-2022] Fix potential memory leak in remote read 

### PR DESCRIPTION
**Pull request by calvinjia**
https://alluxio.atlassian.net/browse/ALLUXIO-2022

### CODE REVIEW

**Comment by calvinjia** [comment]
@pfxuan @gpang @apc999

**Comment by pfxuan** [comment]
@calvinjia Thanks for the patch. I'm going to test it and see if it works!

File: core/server/src/main/java/alluxio/worker/netty/ReleasableResourceChannelListener.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * A listener that will release the given resource when the operation completes. This class accepts
+ * null resources.
+ */
+@ThreadSafe
+final class ReleasableResourceChannelListener implements ChannelFutureListener {
+  private final DataBuffer mResource;
+
+  /**
+   * Creates a new instance of {@link alluxio.worker.netty.ReleasableResourceChannelListener}.
```
Comment (gpang) [inline core/server/src/main/java/alluxio/worker/netty/ReleasableResourceChannelListener.java]: Does this have to be fully specified class name?

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java [hunk 2]
```diff
@@ -52,6 +54,8 @@ public ByteBuffer getReadOnlyByteBuffer() {
 
   @Override
   public void release() {
-    // Nothing we need to release explicitly, let GC take care of all objects.
+    if (mBuffer.isDirect()) {
+      BufferUtils.cleanDirectBuffer(mBuffer);
```
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java]: The assumption is that this is the only owner of this buffer?

**Comment by gpang** [comment]
Thanks @calvinjia ! I left a few questions.

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java (hunk 2 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java]: Yeah, at least we should not call release if we are sharing the buffer (we would need to have some kind of ref count before calling release).

File: core/server/src/main/java/alluxio/worker/netty/ReleasableResourceChannelListener.java (hunk 1 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/worker/netty/ReleasableResourceChannelListener.java]: Updated, thanks.

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java (hunk 2 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java]: can you add a todo for ref count?

Comment (calvinjia) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java]: I don't think we intend to share the reference.

**Comment by pfxuan** [comment]
@calvinjia @gpang @apc999 
GOOD NEWS! I can confirm that the memory leak issue should be fixed after applying this patch. The followings are Worker's JVM usage with two testing cases: 

**Setup:** 
4-node Spark (branch-1.6) co-deployed with Alluxio (branch-1.1) + 8-node OrangeFS (2.9.5). Each compute node has 60GB input datasets. 

**1. Local + UnderFS access:** 

 
 

**2. Local + UnderFS + Remote access:**

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java (hunk 2 above)
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java]: I don't think we have to leave a TODO for ref count, but maybe leave a comment about sharing/not sharing?

**Comment by gpang** [comment]
@calvinjia I left a minor suggestion.

Thanks @pfxuan for the validation!

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java (hunk 2 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataByteBuffer.java]: Updated, thanks.

**Comment by calvinjia** [comment]
@pfxuan Thanks for verifying. To double check, I remember you changed the default to `TRANSFER`, these tests are using this branch (where the default is `MAPPED`) right?

**Comment by pfxuan** [comment]
@calvinjia, the new tests use the default `MAPPED` mode. Also there exists a different issue when switching to `TRANSFER` mode. I'll open another issue for that problem.

In addition, I found the spark job breaks when increasing the size of input from 60 GB to 80 GB per node (Alluxio in-memory size is 62 GB). I don't know if this is related to the new patch. 
   
```Java
 WARN scheduler.TaskSetManager: Lost task 1195.2 in stage 1.0 (TID 1680, 10.125.9.94): org.apache.spark.storage.BlockFetchException: Failed to fetch block from 1 locations. Most recent failure cause:
	at org.apache.spark.storage.BlockManager$$anonfun$doGetRemote$2.apply(BlockManager.scala:605)
	at org.apache.spark.storage.BlockManager$$anonfun$doGetRemote$2.apply(BlockManager.scala:595)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
	at org.apache.spark.storage.BlockManager.doGetRemote(BlockManager.scala:595)
	at org.apache.spark.storage.BlockManager.getRemote(BlockManager.scala:580)
	at org.apache.spark.storage.BlockManager.get(BlockManager.scala:640)
	at org.apache.spark.CacheManager.getOrCompute(CacheManager.scala:44)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:268)
	at org.apache.spark.rdd.ZippedPartitionsRDD2.compute(ZippedPartitionsRDD.scala:88)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:306)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:270)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:306)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:270)
	at org.apache.spark.rdd.PartitionwiseSampledRDD.compute(PartitionwiseSampledRDD.scala:68)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:306)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:270)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:89)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:227)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Connection from /10.125.9.94:42744 closed
	at org.apache.spark.network.client.TransportResponseHandler.channelUnregistered(TransportResponseHandler.java:124)
	at org.apache.spark.network.server.TransportChannelHandler.channelUnregistered(TransportChannelHandler.java:94)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:158)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:144)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:158)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:144)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:158)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:144)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelUnregistered(ChannelInboundHandlerAdapter.java:53)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:158)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelUnregistered(AbstractChannelHandlerContext.java:144)
	at io.netty.channel.DefaultChannelPipeline.fireChannelUnregistered(DefaultChannelPipeline.java:739)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:659)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:357)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
	... 1 more
```

**Comment by pfxuan** [comment]
@calvinjia  Never mind, I think I got the issue from the log file shown in another executor. 

``` Java
WARN server.TransportChannelHandler: Exception in connection from /10.125.9.93:44177
java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3236)
	at java.io.ByteArrayOutputStream.grow(ByteArrayOutputStream.java:118)
	at java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:93)
	at java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:153)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.write(BufferedOutputStream.java:126)
	at com.esotericsoftware.kryo.io.Output.flush(Output.java:155)
	at com.esotericsoftware.kryo.io.Output.require(Output.java:135)
	at com.esotericsoftware.kryo.io.Output.writeLong(Output.java:477)
	at com.esotericsoftware.kryo.io.Output.writeDouble(Output.java:596)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$DoubleArraySerializer.write(DefaultArraySerializers.java:212)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$DoubleArraySerializer.write(DefaultArraySerializers.java:200)
	at com.esotericsoftware.kryo.Kryo.writeObjectOrNull(Kryo.java:549)
	at com.esotericsoftware.kryo.serializers.FieldSerializer$ObjectField.write(FieldSerializer.java:570)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.write(FieldSerializer.java:213)
	at com.esotericsoftware.kryo.Kryo.writeClassAndObject(Kryo.java:568)
	at org.apache.spark.serializer.KryoSerializationStream.writeObject(KryoSerializer.scala:194)
	at org.apache.spark.serializer.SerializationStream.writeAll(Serializer.scala:153)
	at org.apache.spark.storage.BlockManager.dataSerializeStream(BlockManager.scala:1251)
	at org.apache.spark.storage.BlockManager.dataSerialize(BlockManager.scala:1257)
	at org.apache.spark.storage.MemoryStore.getBytes(MemoryStore.scala:195)
	at org.apache.spark.storage.BlockManager.doGetLocal(BlockManager.scala:478)
	at org.apache.spark.storage.BlockManager.getBlockData(BlockManager.scala:300)
	at org.apache.spark.network.netty.NettyBlockRpcServer$$anonfun$2.apply(NettyBlockRpcServer.scala:58)
	at org.apache.spark.network.netty.NettyBlockRpcServer$$anonfun$2.apply(NettyBlockRpcServer.scala:58)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:108)
	at org.apache.spark.network.netty.NettyBlockRpcServer.receive(NettyBlockRpcServer.scala:58)
```

**Comment by pfxuan** [comment]
@calvinjia 
It seams the Spark exception shown above is from Spark 1.6 (branch-1.6). If I run Spark master (branch-2.0), there is no such exception.

**Comment by gpang** [comment]
@pfxuan Thanks for verifying the fix!

**Comment by gpang** [comment]
LGTM

**Comment by apc999** [comment]
+1 LGTM 
thanks @calvinjia  for the fix and thanks for @pfxuan for reporting the issue and verifying the fix.

**Comment by apc999** [comment]
Meanwhile, I am thinking we had a lot of issues with ownership/release java direct buffers before. It seems the best practice of using direct buffer is quite different from a normal java buffer (i.e., using direct buffer is more like a C/C++ buffer where you need to track the ownership, v.s., normal java buffer you just rely on GC), we may want to have a wrapper on top of java direct buffer to make it sort of like those fancy C++11 [smart pointers](http://en.cppreference.com/w/cpp/memory/unique_ptr) with APIs to track ownership.

**Pull request merged**